File: 2025/AHC042/a02.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <chrono>
using namespace std;
 
// 定数
const int N = 20;
 
// グリッド中の 'x'（鬼）の個数を数える関数
int count_x(const vector<string>& grid) {
    int cnt = 0;
    for (int i = 0; i < N; i++){
        for (int j = 0; j < N; j++){
            if (grid[i][j] == 'x')
                cnt++;
        }
    }
    return cnt;
}
 
// グリッド中の 'o'（福）の個数を数える関数
int count_o(const vector<string>& grid) {
    int cnt = 0;
    for (int i = 0; i < N; i++){
        for (int j = 0; j < N; j++){
            if (grid[i][j] == 'o')
                cnt++;
        }
    }
    return cnt;
}
 
// 盤面のコピーを作成し、指定した行または列を指定方向にシフトする。
// d: 移動方向 ('L', 'R', 'U', 'D')
// p: 移動対象の行（d=='L'または'R'の場合）または列（d=='U'または'D'の場合）のインデックス
vector<string> move_grid(const vector<string>& grid, char d, int p) {
    vector<string> new_grid = grid; // ディープコピー
    if (d == 'L') {
        // 行 p を左にシフト
        for (int i = 0; i < N - 1; i++) {
            new_grid[p][i] = grid[p][i+1];
        }
        new_grid[p][N-1] = '.';
    } else if (d == 'R') {
        // 行 p を右にシフト
        new_grid[p][0] = '.';
        for (int i = 1; i < N; i++) {
            new_grid[p][i] = grid[p][i-1];
        }
    } else if (d == 'U') {
        // 列 p を上にシフト
        for (int i = 0; i < N - 1; i++) {
            new_grid[i][p] = grid[i+1][p];
        }
        new_grid[N-1][p] = '.';
    } else if (d == 'D') {
        // 列 p を下にシフト
        new_grid[0][p] = '.';
        for (int i = 1; i < N; i++) {
            new_grid[i][p] = grid[i-1][p];
        }
    }
    return new_grid;
}
 
// 指定位置 (i,j) にある鬼（'x'）について、上下左右それぞれの方向で
// 鬼が落ちるまでの距離の 3 乗の逆数を計算し、その中で最大のものを返す。
// 途中に福（'o'）があれば、その方向は無視する。
double calc_x_eval(const vector<string>& grid, int i, int j) {
    double bestVal = 0.0;
    vector<pair<int,int>> dirs = { {0,1}, {0,-1}, {1,0}, {-1,0} };
    for (auto &dir : dirs) {
        int di = dir.first, dj = dir.second;
        int x = i, y = j;
        bool blocked = false;
        int dist = 1;
        x += di; y += dj;
        while (x >= 0 && x < N && y >= 0 && y < N) {
            if (grid[x][y] == 'o') {
                blocked = true;
                break;
            }
            if (grid[x][y] == '.')
                dist++;
            x += di; y += dj;
        }
        if (blocked) continue;
        double candidate = 1.0 / (dist * dist * dist);
        bestVal = max(bestVal, candidate);
    }
    return bestVal;
}
 
// 盤面全体の評価値を計算する関数
double calc_eval_01(const vector<string>& grid) {
    double ev = 0.0;
    // 福を落とした場合の大きなペナルティ
    ev -= 1e9 * (2 * N - count_o(grid));
    // 鬼を落とした場合の報酬
    ev += 1e5 * (2 * N - count_x(grid));
    // 各鬼について、その鬼が落ちるまでの評価を加算
    for (int i = 0; i < N; i++){
        for (int j = 0; j < N; j++){
            if (grid[i][j] == 'x'){
                ev += calc_x_eval(grid, i, j);
            }
        }
    }
    return ev;
}
 
// 探索状態を表す構造体
struct State {
    bool visited;                       // すでに展開済みか
    double eval;                        // 評価値
    vector<string> grid;                // 現在の盤面
    vector<pair<char,int>> moves;       // これまでの操作の履歴 (方向, 行または列のインデックス)
};
 
// chokudaiサーチ（ビームサーチ）の実装
// 盤面と開始時刻を受け取り、全ての鬼を落とす操作列を返す（見つからなければ空のリスト）。
vector<pair<char,int>> serch_02(const vector<string>& initGrid,
    chrono::steady_clock::time_point start_time) {
    
    vector<pair<char,int>> shortestAns; // 最短の操作列
    int min_len = 4 * N * N + 1;          // 許容する最大操作回数 (4*N^2+1)
    // 各操作回数ごとの状態リスト
    vector<vector<State>> beam(4 * N * N + 1);
 
    double initEval = calc_eval_01(initGrid);
    State initState;
    initState.visited = false;
    initState.eval = initEval;
    initState.grid = initGrid;
    // 操作履歴は空で初期化
    beam[0].push_back(initState);
 
    // 実行時間が約1.9秒以内で探索する
    while (true) {
        auto now = chrono::steady_clock::now();
        double elapsed = chrono::duration<double>(now - start_time).count();
        cerr << "Elapsed time: " << elapsed << " seconds" << endl;
        if (elapsed >= 1.9)
            break;
 
        bool found = false;
        // 各操作回数（深さ）について探索
        for (int i = 0; i < min_len - 1; i++){
            if (beam[i].empty()) continue;
            // すでにこの深さの先頭状態が展開済みならスキップ
            if (beam[i][0].visited)
                continue;
 
            State cur = beam[i][0]; // 最高評価の状態を選択
 
            // 4方向すべてについて
            for (char d : {'L', 'R', 'U', 'D'}) {
                if (found) break;
                // すべての行または列 (0～N-1) に対して操作を試す
                for (int p = 0; p < N; p++){
                    vector<string> newGrid = move_grid(cur.grid, d, p);
                    double newEval = calc_eval_01(newGrid);
                    vector<pair<char,int>> newMoves = cur.moves;
                    newMoves.push_back({d, p});
 
                    State newState;
                    newState.visited = false;
                    newState.eval = newEval;
                    newState.grid = newGrid;
                    newState.moves = newMoves;
 
                    beam[i+1].push_back(newState);
 
                    // もし盤面上に鬼がなくなったなら解を記録
                    if (count_x(newGrid) == 0) {
                        min_len = i + 1;
                        shortestAns = newMoves;
                        found = true;
                        break;
                    }
                }
            }
            // 次の深さの状態リストを、未展開かつ評価値の高い順にソート
            sort(beam[i+1].begin(), beam[i+1].end(), [](const State &a, const State &b) {
                if (a.visited != b.visited)
                    return a.visited < b.visited; // false (未展開) が先
                return a.eval > b.eval; // 評価値の高い順
            });
            // 現在の状態を展開済みとする
            beam[i][0].visited = true;
            if (found)
                break;
        }
        if (!shortestAns.empty())
            break;
    }
    return shortestAns;
}
 
int main(){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
 
    // 入力
    // 最初の整数はグリッドのサイズ（ここでは20と想定）
    int n;
    cin >> n;
    vector<string> grid(n);
    for (int i = 0; i < n; i++){
        cin >> grid[i];
    }
 
    auto start_time = chrono::steady_clock::now();
    // chokudaiサーチで操作列を探索
    vector<pair<char,int>> ans = serch_02(grid, start_time);
 
    // 出力：各行に "方向 インデックス" を出力
    for (auto &mv : ans){
        cout << mv.first << " " << mv.second << "\n";
    }
 
    return 0;
}

File: 2025/AHC042/a02.py
import os
import time

# デバッグ用  気にしなくていい
MyPC = os.path.basename(__file__) != "Main.py"
if MyPC:
  from icecream import ic
  ic.disable()
else:
  def ic(*args):
    return None

ic.enable() if MyPC else None

# 概要
# 20x20のグリッドが与えられる
# このグリッドの中に、"o" と "x" がある
# "o" は福、"x" は鬼を表す
# 行または列をベルトコンベアのように動かし、鬼をすべて落とす
# ただし、福を落としてはいけない
# 操作の回数は少ないほど良い
# その時の動作を出力する

N = 20

# 鬼を数える
def count_x(grid):
  count = 0
  for i in range(N):
    for j in range(N):
      if grid[i][j] == "x":
        count += 1
  return count

# 福を数える
def count_o(grid):
  count = 0
  for i in range(N):
    for j in range(N):
      if grid[i][j] == "o":
        count += 1
  return count

# 1行または1列を動かした時の結果を返す
def move(grid, d, p):  # grid: 盤面, d: 方向, p: 位置
  new_grid = [row[:] for row in grid]  # deepcopy
  # 左方向の場合
  if d == "L":
    for i in range(N-1):
      new_grid[p][i] = grid[p][i+1]
    new_grid[p][N-1] = "."  # 一番右の列は空白にする
  # 右方向の場合
  elif d == "R":
    new_grid[p][0] = "."  # 一番左の列は空白にする
    for i in range(1, N):
      new_grid[p][i] = grid[p][i-1]
  # 上方向の場合
  elif d == "U":
    for i in range(N-1):
      new_grid[i][p] = grid[i+1][p]
    new_grid[N-1][p] = "."  # 一番下の行は空白にする
  # 下方向の場合
  elif d == "D":
    new_grid[0][p] = "."  # 一番上の行は空白にする
    for i in range(1, N):
      new_grid[i][p] = grid[i-1][p]
  return new_grid

# 鬼が落ちるまでの距離の3乗の逆数を計算
def calc_x_eval(grid, i, j):
  eval = 0
  # 上下左右の4方向に対して
  for di, dj in [(0, 1), (0, -1), (1, 0), (-1, 0)]:
    x = i
    y = j
    flag = False
    dist = 1
    x += di
    y += dj
    # 鬼が落ちるまでの距離を計算
    while 0 <= x < N and 0 <= y < N:
      if grid[x][y] == "o":
        flag = True
        break
      if grid[x][y] == ".":
        dist += 1
      x += di
      y += dj
    # 福がある場合はその経路は通らない
    if flag:
      continue
    # 評価値を計算し、大きい方を採用
    eval = max(eval, 1/dist**3)
  return eval

# 評価値を計算する
def calc_eval_01(grid):
  eval = 0
  # 福を落とした場合、-10**9
  eval -= 10**9 * (2*N - count_o(grid))
  # 鬼を落とした場合、+10**5
  eval += 10**5 * (2*N - count_x(grid))
  # 各鬼に対し、その鬼が落ちるまでの距離を計算
  # 距離の3乗の逆数を評価値に加算
  # ただし、福がある場合はその経路は通らない
  for i in range(N):
    for j in range(N):
      if grid[i][j] == "x":
        eval += calc_x_eval(grid, i, j)
  return eval

# 貪欲法で解く  # もう使わない関数
# def serch_01(grid, ans_list):
#   max_eval = -10**18
#   max_grid = []
#   max_move = ()
#   # 1行または1列を動かす
#   for d in ["L", "R", "U", "D"]:
#     for p in range(N):
#       new_grid = move(grid, d, p)
#       # 評価値を計算
#       eval = calc_eval_01(new_grid)
#       # 評価値が最大の時
#       if max_eval < eval:
#         max_eval = eval
#         max_grid = new_grid
#         max_move = (d, p)
#   # 結果を保存
#   ans_list.append(max_move)
#   # 鬼が1つもなくなったら終了
#   if count_x(max_grid) == 0:
#     return ans_list
#   # 操作回数が4*N^2回になっても終了
#   elif len(ans_list) == 4*N**2:
#     return ans_list
#   # 次の手を探す
#   return serch_01(max_grid, ans_list)

# chokudaiサーチで解く
# 再起のやり方がわからないので、for文で書く
def serch_02(grid, start_time):
  shortest_ans_list = []  # 最短の操作
  min_len = 4*N**2+1  # 最短の操作回数
  chokudai_list = [[] for _ in range(4*N**2+1)]  # chokudaiサーチのための、盤面のリスト。操作回数ごとに保存
  # 最初の盤面について、操作回数0のリストに追加
  # その際、評価値とそれまでの操作と探索済みかどうかを保存
  # [探索済みか, 評価値, 盤面, 操作] の順
  eval = calc_eval_01(grid)
  chokudai_list[0].append([False, eval, grid, []])
  # 実行時間が1.9秒以内である限りビームを打ち続ける
  while time.time() - start_time < 1.9:
    # 操作回数ごとに探索
    # 操作回数が最も少ないものを探すため、すでに見つかっている最短の操作回数より多い場合は探索を終了
    for i in range(min_len-1):
      flag = False  # 最短の操作が見つかったかどうか
      # 探索済みでない盤面について
      # 探索済みでない盤面のうち、最も評価値が高いものを探索
      if chokudai_list[i][0][0]:
        continue
      # すべての操作について、次の盤面を探索
      for d in ["L", "R", "U", "D"]:
        # 最短の操作がすでに見つかっている場合、探索を終了
        if flag:
          break
        for p in range(N):
          new_grid = move(chokudai_list[i][0][2], d, p)
          # 評価値を計算
          eval = calc_eval_01(new_grid)
          # ここまでの操作を記録
          operation_list = chokudai_list[i][0][3][:] + [(d, p)]
          # chokudai_listに追加
          chokudai_list[i+1].append([False, eval, new_grid, operation_list])
          # 鬼がすべて落ちた場合、ビームを打ち切る
          if count_x(new_grid) == 0:
            # 最短の操作回数を更新
            min_len = i+1
            shortest_ans_list = operation_list
            flag = True
            break
      # 今追加した階層について、未探索のものを優先して、評価値が高い順にソート
      chokudai_list[i+1].sort(key=lambda x: (x[0], -x[1]))
      # 探索済みにする
      chokudai_list[i][0][0] = True
      # 最短の操作がすでに見つかっている場合、探索を終了
      if flag:
        break
  # 結果を返す
  return shortest_ans_list

def main():
  start_time = time.time()  # 開始時刻
  n = int(input())  # 20固定
  grid = [[""]*N for _ in range(N)]  # 20x20のグリッド
  # 20行分の入力
  # "o": 福, "x": 鬼, ".": 何もない
  for i in range(N):
    s = input()  # "o", "x", "." からなる文字列
    for j in range(N):
      grid[i][j] = s[j]

  # chokudaiサーチで解く
  ans_list = serch_02(grid, start_time)
  # 出力
  # print(len(ans_list))
  for ans in ans_list:
    print(*ans)

if __name__ == "__main__":
  main()